File: src/onedimensiondp/lc_238_product_array_except_self/ProductArrayExceptSelf.java

class ProductArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] res = new int[length];
        Arrays.fill(res, 1);
        int shift = 1;
        for(int i = 0; i < length; i++){
            res[i] *= shift;
            shift *= nums[i];
        }
        shift = 1;
        for(int i = length - 1; i >= 0; i--){
            res[i] *= shift;
            shift *= nums[i];
        }
        return res;
    }
}

File: src/onedimensiondp/lc_238_product_array_except_self/product_array_except_self.py
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        length = len(nums)
        res = [1] * length

        pre = 1
        for i in range(length):
            res[i] *= pre
            pre *= nums[i]
        
        # all products except with the last one
        post = 1
        for i in range(length -1,-1,-1):
            res[i] *= post
            post *= nums[i] #multiply all except self
        return res